File: array/problem46/problem46.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        # Backtracking by fixing one position at a time using in-place swaps
        res: List[List[int]] = []
        n = len(nums)

        def dfs(first: int) -> None:
            if first == n:
                res.append(nums[:])
                return
            for i in range(first, n):
                nums[first], nums[i] = nums[i], nums[first]
                dfs(first + 1)
                nums[first], nums[i] = nums[i], nums[first]

        dfs(0)
        return res

File: array/problem46/problem46.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        // Backtracking with in-place swaps to generate all n! permutations
        vector<vector<int>> res;
        dfs(0, nums, res);
        return res;
    }

private:
    void dfs(int first, vector<int>& nums, vector<vector<int>>& res) {
        if (first == static_cast<int>(nums.size())) {
            res.push_back(nums);
            return;
        }
        for (int i = first; i < static_cast<int>(nums.size()); ++i) {
            swap(nums[first], nums[i]);
            dfs(first + 1, nums, res);
            swap(nums[first], nums[i]);
        }
    }
};
